# SVM Classification

This data set is made available by http://www.baseball-reference.com. It contains data about professional baseball (MLB) games played in the 2016 season. There are 2,427 games in the data set. Each row represents a single game. Predict the attendance at a home team’s game. This is an important task because most franchises want to predict the number of attendees for a variety of reasons including profits.

## Goal

Use the **baseball.csv** data set and build a model to predict **attendance_binary**. 

# Read and Prepare the Data:

## Import Data

In [23]:
import numpy as np
import pandas as pd

np.random.seed(42)

In [24]:
baseball = pd.read_csv("baseball.csv")

#Drop GameID column since it can't be used as a feature
baseball = baseball.drop(['GameID'], axis=1)
baseball.head()

attendance_binary  previous_attendance  previous_away_team_errors  \
0                  0                43683                          2   
1                  0                45785                          0   
2                  0                48282                          0   
3                  0                21830                          0   
4                  0                49289                          2   

   previous_away_team_hits  previous_away_team_runs   game_type  \
0                        6                        2  Night Game   
1                        7                        2  Night Game   
2                        8                        4  Night Game   
3                        9                        6    Day Game   
4                        4                        2  Night Game   

  previous_game_type  previous_home_team_errors  previous_home_team_hits  \
0           Day Game                          0                        6   
1           Day Game                          0                       10   
2           Day Game                          2                        4   
3         Night Game                          0                       15   
4           Day Game                          1                        1   

   previous_home_team_runs   game_day previous_game_day  temperature  \
0                        6  Wednesday            Monday           55   
1                        3  Wednesday            Monday           48   
2                        3  Wednesday            Monday           65   
3                       11  Wednesday           Tuesday           77   
4                        3    Tuesday            Monday           81   

   wind_speed       sky  previous_game_duration  previous_homewin  
0          24  Overcast                2.933333                 1  
1           7   Unknown                2.800000                 1  
2          10    Cloudy                3.383333                 0  
3           0   In Dome                3.233333                 1  
4          12    Cloudy                2.633333                 1

## Check for Missing Values

In [25]:
baseball.isna().sum()

attendance_binary            0
previous_attendance          0
previous_away_team_errors    0
previous_away_team_hits      0
previous_away_team_runs      0
game_type                    0
previous_game_type           0
previous_home_team_errors    0
previous_home_team_hits      0
previous_home_team_runs      0
game_day                     0
previous_game_day            0
temperature                  0
wind_speed                   0
sky                          0
previous_game_duration       0
previous_homewin             0
dtype: int64

## Split Data

In [26]:
from sklearn.model_selection import train_test_split

train, test = train_test_split(baseball, test_size=0.3)

## Separate Target

In [27]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

train_y = train[['attendance_binary']]
test_y = test[['attendance_binary']]

train_inputs = train.drop(['attendance_binary'], axis=1)
test_inputs = test.drop(['attendance_binary'], axis=1)

## Identify Numeric and Categorical Columns

In [28]:
train_inputs.dtypes

previous_attendance            int64
previous_away_team_errors      int64
previous_away_team_hits        int64
previous_away_team_runs        int64
game_type                     object
previous_game_type            object
previous_home_team_errors      int64
previous_home_team_hits        int64
previous_home_team_runs        int64
game_day                      object
previous_game_day             object
temperature                    int64
wind_speed                     int64
sky                           object
previous_game_duration       float64
previous_homewin               int64
dtype: object

In [29]:
# Identify the numeric columns
numeric_columns = train_inputs.select_dtypes(include=[np.number]).columns
numeric_columns

Index(['previous_attendance', 'previous_away_team_errors',
       'previous_away_team_hits', 'previous_away_team_runs',
       'previous_home_team_errors', 'previous_home_team_hits',
       'previous_home_team_runs', 'temperature', 'wind_speed',
       'previous_game_duration', 'previous_homewin'],
      dtype='object')

In [30]:
# Identify the categorical columns
categorical_columns = train_inputs.select_dtypes('object').columns
categorical_columns

Index(['game_type', 'previous_game_type', 'game_day', 'previous_game_day',
       'sky'],
      dtype='object')

## Pipeline

In [31]:
numeric_transformer = Pipeline(steps=[               
                ('scaler', StandardScaler())])

categorical_transformer = Pipeline(steps=[  
    ('onehot', OneHotEncoder(handle_unknown='ignore'))])

preprocessor = ColumnTransformer([
        ('num', numeric_transformer, numeric_columns),
        ('cat', categorical_transformer, categorical_columns)])

## Transform Train and Test

In [32]:
#Fit and transform the train data
train_x = preprocessor.fit_transform(train_inputs)
train_x.shape

(1698, 37)

In [33]:
# Transform the test data
test_x = preprocessor.transform(test_inputs)
test_x.shape

(729, 37)

# Determine Baseline Accuracy

In [34]:
# Find majority class
train_y['attendance_binary'].value_counts()

1    873
0    825
Name: attendance_binary, dtype: int64

In [35]:
# Find percentage
train_y['attendance_binary'].value_counts()/len(train_y)

1    0.514134
0    0.485866
Name: attendance_binary, dtype: float64

A naive prediction of 1 yields an accuracy of only 51.4%

# SVM Model 1 - SVC(kernel=linear)

In [41]:
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC
 
lin_svm = SVC(kernel="linear")

lin_svm.fit(train_x, train_y)

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='linear', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)

In [42]:
#Predict the train values
train_y_pred = lin_svm.predict(train_x)

#Train accuracy
accuracy_score(train_y, train_y_pred)

0.8445229681978799

In [43]:
#Predict the test values
test_y_pred = lin_svm.predict(test_x)

#Test accuracy
accuracy_score(test_y, test_y_pred)

0.8161865569272977

A difference in accuracy of fewer than 3 percentage points indicates that there is not much overfitting.

# SVM Model 2 - SVC(kernel=poly)

In [81]:
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC
 
pol_svm = SVC(kernel="poly", degree=2, coef0=1, C=5, gamma='scale')

pol_svm.fit(train_x, train_y)

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(C=5, cache_size=200, class_weight=None, coef0=1,
    decision_function_shape='ovr', degree=2, gamma='scale', kernel='poly',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [82]:
#Predict the train values
train_y_pred = pol_svm.predict(train_x)

#Train accuracy
accuracy_score(train_y, train_y_pred)

0.8957597173144877

In [83]:
#Predict the test values
test_y_pred = pol_svm.predict(test_x)

#Test accuracy
accuracy_score(test_y, test_y_pred)

0.8079561042524005

The degree-2 polynomial model performed slightly worse than the linear model, and had noticeably worse overfitting (~9 percentage points).

# SVM Model 3 - SVC(kernel=rbf)

In [75]:
rbf_svm = SVC(kernel="rbf", C=10, gamma=0.1)

rbf_svm.fit(train_x, train_y)

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(C=10, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.1, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [76]:
#Predict the train values
train_y_pred = rbf_svm.predict(train_x)

#Train accuracy
accuracy_score(train_y, train_y_pred)

0.9929328621908127

In [77]:
#Predict the test values
test_y_pred = rbf_svm.predict(test_x)

#Test accuracy
accuracy_score(test_y, test_y_pred)

0.7983539094650206

A much larger difference in accuracy of nearly 20 percentage points indicates significant overfitting.

# SVM Model 4 - SVC(rbf) with Grid Search

In [90]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint, uniform
import random

#define ranges for C and gamma
param_distribs = {
        'C': randint(low=5, high=70),
        'gamma': uniform(0.01, 0.5),    
    }

#define model
rbf_svm = SVC(kernel="rbf")

#search variations of hyperparameters 50 iterations
rbf_search = RandomizedSearchCV(rbf_svm, param_distributions=param_distribs,
                                n_iter=50, cv=10, scoring='accuracy', random_state=42)

#fit models on training data
rbf_search.fit(train_x, train_y)

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\ajohnson\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning

RandomizedSearchCV(cv=10, error_score='raise-deprecating',
                   estimator=SVC(C=1.0, cache_size=200, class_weight=None,
                                 coef0=0.0, decision_function_shape='ovr',
                                 degree=3, gamma='auto_deprecated',
                                 kernel='rbf', max_iter=-1, probability=False,
                                 random_state=None, shrinking=True, tol=0.001,
                                 verbose=False),
                   iid='warn', n_iter=50, n_jobs=None,
                   param_distributions={'C': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001DE1B8F8F48>,
                                        'gamma': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001DE1C962888>},
                   pre_dispatch='2*n_jobs', random_state=42, refit=True,
                   return_train_score=False, scoring='accuracy', verbose=0)

In [91]:
pd.DataFrame(rbf_search.cv_results_).sort_values('mean_test_score',ascending=False)

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
36       0.069216      0.001424         0.005086    2.995339e-04      12   
18       0.073703      0.000946         0.004986    3.996624e-07      13   
7        0.080885      0.001437         0.004988    6.687801e-06      26   
21       0.083881      0.004304         0.005480    4.867632e-04      12   
16       0.101624      0.003498         0.005388    4.923228e-04      43   
19       0.139229      0.002899         0.011071    2.927390e-04       6   
38       0.138132      0.003552         0.011471    1.198506e-03      12   
2        0.140119      0.002752         0.011076    5.437079e-04      28   
47       0.138231      0.002414         0.011370    4.885491e-04       5   
39       0.138627      0.001733         0.010872    2.996989e-04      37   
49       0.137632      0.002318         0.011370    4.886945e-04      13   
1        0.141124      0.004044         0.012663    5.426647e-03      65   
31       0.137530      0.000828         0.010771    3.991485e-04      57   
44       0.138431      0.001244         0.010871    2.994937e-04      48   
35       0.137233      0.001558         0.011569    6.616523e-04      13   
9        0.123167      0.002961         0.005691    4.575958e-04      46   
46       0.137034      0.000914         0.011470    4.988671e-04      51   
32       0.139330      0.002001         0.010472    4.990343e-04      64   
8        0.140327      0.004899         0.010770    1.163483e-03      53   
17       0.138234      0.001629         0.010668    4.555301e-04      64   
48       0.137532      0.001916         0.009874    2.992789e-04      18   
45       0.138632      0.003706         0.009772    3.978661e-04      39   
26       0.135937      0.001342         0.011569    4.887724e-04      67   
23       0.132845      0.002593         0.008278    4.573245e-04       8   
14       0.136537      0.004521         0.011865    5.305637e-04       7   
10       0.132243      0.001907         0.008579    4.902596e-04      64   
42       0.134242      0.002862         0.008876    5.376410e-04      52   
41       0.133740      0.001916         0.008378    4.891035e-04      16   
29       0.133743      0.004937         0.008178    3.989936e-04      52   
43       0.133144      0.002499         0.008079    2.990544e-04      66   
25       0.125266      0.003518         0.007081    2.992711e-04       8   
30       0.132449      0.001535         0.008082    2.981968e-04      66   
6        0.131746      0.001691         0.007878    2.991538e-04      37   
33       0.132249      0.001800         0.008079    2.991796e-04      45   
40       0.135239      0.001679         0.011772    4.003102e-04      45   
24       0.128457      0.002266         0.006981    4.461150e-04      10   
28       0.133943      0.000779         0.011469    4.987718e-04      66   
27       0.134840      0.001163         0.011669    4.571631e-04      48   
15       0.132474      0.011318         0.007089    6.957889e-04      11   
0        0.135633      0.002613         0.011676    4.566236e-04      56   
37       0.129753      0.002495         0.006982    4.446771e-04      15   
3        0.134743      0.001133         0.011369    4.887516e-04      57   
34       0.125365      0.004506         0.005489    5.023730e-04      49   
13       0.135343      0.002287         0.011666    4.545035e-04      59   
22       0.133447      0.002633         0.006784    3.935209e-04      40   
5        0.135440      0.002216         0.011471    5.004760e-04      68   
4        0.129458      0.001316         0.006778    3.989028e-04      34   
12       0.131955      0.003185         0.006977    4.430292e-04      66   
20       0.135608      0.017672         0.006378    4.906102e-04      64   
11       0.123275      0.003437         0.005879    2.997935e-04      66   

   param_gamma                                    params  split0_test_score  \
36   0.0127611    {'C': 12, 'gamma': 0.012761058

## Determine the accuracy of the best model

In [96]:
#Pick best model and display
final_model = rbf_search.best_estimator_
rbf_search.best_params_

{'C': 12, 'gamma': 0.0127610585618012}

In [97]:
#Test the model accuracy on the train set
train_predictions = final_model.predict(train_x)

accuracy_score(train_y, train_predictions)

0.8769140164899882

In [98]:
#Test the model accuracy on the test set
test_predictions = final_model.predict(test_x)

accuracy_score(test_y, test_predictions)

0.8161865569272977

The grid search choices of a higher C value and lower gamma gave better results.

The best performing model of the four seems to be the first, simple linear model. The test accuracy of that model is 81.6% and had only 3 percentage points of separation from the train accuracy, indicating low overfitting. In contrast, the polynomial model only achieved 80.8% test accuracy and had noticeable overfitting with a 9 percentage point difference to the train accuracy. The first rbf model failed to improve on the linear model, and had major overfitting. To attempt to find more suitable hyperparameters and allow for a lower gamma (to address overfitting), a randomized grid search found a model that yielded a test accuracy of 81.6%; however this still fails to meaningfully improve on the accuracy original linear model, and still has more overfitting.